### PR TITLE
[Feature] [Bugfix] [Refactor] Fix shadows and refactor measurement protocols

### LINKED ISSUE #39: [Enhancement] Shadows's compute_traces reduce bottleneck
In compute_traces, we reshape the observable matrix to fit the density matrix dimensions by filling identities. We should avoid doing so to reduce the computational cost, and make this faster.

### LINKED ISSUE #48: [Refactoring] Refactor the measurements for better usage and to use the new NoiseHandler interface from Qadence
Refactor the current code to introduce a class per measurement protocol that keeps as an attribute the measurement data which can be reused to compute other quantities of interest. Also, this should make the code clearer without blotting functions with many input parameters. 

Finally, we can update the restriction on the Qadence version and use the `NoiseHandler` interface introduced [here](https://github.com/pasqal-io/qadence/pull/591).

### PR DESCRIPTION
**Pull request by chMoussa**
Closes #39, #48 and #47.

We introduce the concept of `MeasurementManager` to keep in memory the measurement data that can be reused to reconstruct states or calculate over quantities of interests. We also separate each functionality into a different class method for the sake of clarity and we improved the speed of execution.

### CODE REVIEW

File: docs/measurements/index.md [hunk 1]
```diff
@@ -85,23 +96,60 @@ from qadence_protocols.measurements.utils_shadow import number_of_samples
...
 ```
 
+## Getting shadows
+
+If we are interested in accessing the measurement data from shadows, we can access the measurement data via the `measurement_manager` attribute or simply pass `return_expectations=False` (a new round of measurements is performed and returned directly instead of computing expectation values) as follows:
+
+```python exec="on" source="material-block" session="measurements" result="json"
+
+measurements_shadows = shadow_measurement.measurement_manager.measurement_data
+# measurements_shadows = shadow_measurement(model, return_expectations=False)
```
Comment (RolandMacDoland) [inline docs/measurements/index.md]: Remove ?

File: docs/measurements/index.md [hunk 2]
```diff
@@ -85,23 +96,60 @@ from qadence_protocols.measurements.utils_shadow import number_of_samples
...
 print(f"Estimated expectation value shadow = {estimated_values_shadow}") # markdown-exec: hide
 ```
 
+## Getting shadows
+
+If we are interested in accessing the measurement data from shadows, we can access the measurement data via the `measurement_manager` attribute or simply pass `return_expectations=False` (a new round of measurements is performed and returned directly instead of computing expectation values) as follows:
+
+```python exec="on" source="material-block" session="measurements" result="json"
+
+measurements_shadows = shadow_measurement.measurement_manager.measurement_data
```
Comment (RolandMacDoland) [inline docs/measurements/index.md]: I would simplify as much as possible: `shadow_measurement.manager.data`

File: docs/measurements/index.md [hunk 3]
```diff
@@ -85,23 +96,60 @@ from qadence_protocols.measurements.utils_shadow import number_of_samples
...
+# measurements_shadows = shadow_measurement(model, return_expectations=False)
+
+print("Sampled unitary indices shape: ", measurements_shadows[0].shape)
+print("Shape of batched measurements: ", measurements_shadows[1].shape)
+```
+
+In the case of shadows, the measurement data is a tuple of two elements:
```
Comment (RolandMacDoland) [inline docs/measurements/index.md]: If that's a pair of elements, I'd give them attributes to access instead of indexing.

File: qadence_protocols/measurements/abstract.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+class MeasurementManager(ABC):
+    """The abstract class that defines the interface for the managing measurements.
+
+    Attributes:
+        measurement_data (MeasurementData, optional): Measurement data if already obtained.
+        options (dict, optional): Dictionary of options specific to protocol.
+    """
+
+    def __init__(self, measurement_data: MeasurementData = None, options: dict = dict()):
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/abstract.py]:
```suggestion
    def __init__(self, measurement_data: MeasurementData | None = None, options: dict = dict()):
```

Unless `None` makes sense wrt `MeasurementData` type.

File: qadence_protocols/measurements/abstract.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def verify_options(self) -> dict:
+        """Validate options passed to procotol."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def reconstruct_state(self, snapshots: Tensor) -> Tensor:
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/abstract.py]: I guess you can make this method depend on the next one and you don't have to pass `snapshots`.

File: qadence_protocols/measurements/abstract.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+            snapshots (Tensor): Snapshots of size
+                (batch_size, shadow_size, 2**n, 2**n).
+
+        Returns:
+            Tensor: Reconstructed state.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_snapshots(
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/abstract.py]:
```suggestion
    def snapshots(
```

File: qadence_protocols/measurements/robust_shadow.py [hunk 7]
```diff
@@ -1,101 +1,151 @@
...
+
+
+class RobustShadowManager(MeasurementManager):
+    """The class for managing randomized robust shadow."""
+
+    def __init__(self, measurement_data: MeasurementData = None, options: dict = dict()):
+        self.measurement_data = measurement_data
+        self.options = options
+
+    def verify_options(self) -> dict:
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/robust_shadow.py]: Not sure I quite get what this method means. From the name it feels like it should validate the options. If that's the case, it should return a `bool` ?

File: qadence_protocols/measurements/shadow.py [hunk 8]
```diff
@@ -4,90 +4,159 @@
...
+        confidence = self.options.get("confidence", None)
+        if confidence is None:
+            raise KeyError("Shadow protocol requires an option 'confidence' of type 'float'.")
+        self.options = {"shadow_size": shadow_size, "accuracy": accuracy, "confidence": confidence}
+        return self.options
+
+    def number_of_samples_and_groups(
+        self, observables: list[AbstractBlock] = list()
+    ) -> tuple[int, ...]:
+        """Return the shadow size and number of groups for a list of observables.
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/shadow.py]: What is the number of groups here ?

File: qadence_protocols/measurements/shadow.py [hunk 9]
```diff
@@ -4,90 +4,159 @@
...
+            state (Tensor | None, optional): Input state. Defaults to None.
+
+        Returns:
+            MeasurementData: Measurement data as locally sampled pauli unitaries and
+                samples from the circuit
+                rotated according to the locally sampled pauli unitaries.
+        """
+
+        circuit = model._circuit.original
+        shadow_size = self.options["shadow_size"]
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/shadow.py:161]: Have you verified these options are present here ?

File: qadence_protocols/measurements/tomography.py [hunk 10]
```diff
@@ -6,89 +6,110 @@
...
+    def reconstruct_state(self, snapshots: Tensor) -> Tensor:
+        raise NotImplementedError
+
+    def get_snapshots(
+        self,
+        model: QuantumModel,
+        param_values: dict[str, Tensor] = dict(),
+        state: Tensor | None = None,
+    ) -> Tensor:
+        raise NotImplementedError
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/tomography.py]: I think it would be better not to define them here if possible. Maybe create `ShadowManager` subclass that would hold these abstract methods ?

File: qadence_protocols/types.py [hunk 11]
```diff
@@ -1,6 +1,11 @@
 from __future__ import annotations
 
 from enum import Enum
+from typing import Union
+
+from torch import Tensor
+
+MeasurementData = Union[Tensor, list, tuple, None]
```
Comment (RolandMacDoland) [inline qadence_protocols/types.py]:
```suggestion
MeasurementData = Tensor | list | tuple | None
```

For a reason, I would leave `None` out.

**Review by RolandMacDoland — COMMENTED** [review]
Thanks a lot @chMoussa. Looking good. Some comments about general simplification. Feel free to discuss here.

File: qadence_protocols/measurements/abstract.py (hunk 4 above)
Comment (chMoussa) [reply qadence_protocols/measurements/abstract.py]: For me None refers to no data obtained yet

File: qadence_protocols/measurements/robust_shadow.py (hunk 7 above)
Comment (chMoussa) [reply qadence_protocols/measurements/robust_shadow.py]: More like it checks the options are right and raises an error if not. It also change the dict internally with the default values if they were not passed by the user.

File: qadence_protocols/measurements/shadow.py (hunk 8 above)
Comment (chMoussa) [reply qadence_protocols/measurements/shadow.py]: It is K, which is how many groups of measurements you use within the median of means algo part. Maybe I should call it differently. How about `shadow_medians` and `number_of_samples_and_medians`?

File: qadence_protocols/measurements/shadow.py (hunk 9 above)
Comment (chMoussa) [reply qadence_protocols/measurements/shadow.py:161]: Yes the `verify_options` takes care of it. Maybe we can call it `validate_options` for clarity?

File: qadence_protocols/measurements/tomography.py (hunk 10 above)
Comment (chMoussa) [reply qadence_protocols/measurements/tomography.py]: Ok, let me indeed separate them.

File: docs/measurements/index.md (hunk 3 above)
Comment (chMoussa) [reply docs/measurements/index.md]: Ok, maybe I shall consider giving back a dictionary instead of a tuple.

File: qadence_protocols/measurements/robust_shadow.py (hunk 7 above)
Comment (chMoussa) [reply qadence_protocols/measurements/robust_shadow.py]: To call `validate_options` maybe?

File: qadence_protocols/measurements/abstract.py (hunk 4 above)
Comment (chMoussa) [reply qadence_protocols/measurements/abstract.py]: Ah ok, you mean decouple the None from measurement data

File: qadence_protocols/measurements/tomography.py (hunk 10 above)
Comment (chMoussa) [reply qadence_protocols/measurements/tomography.py]: Done

File: qadence_protocols/types.py (hunk 11 above)
Comment (chMoussa) [reply qadence_protocols/types.py]: Still have to use Union for python 3.9

File: qadence_protocols/measurements/abstract.py (hunk 5 above)
Comment (chMoussa) [reply qadence_protocols/measurements/abstract.py]: Changed

File: docs/measurements/index.md (hunk 3 above)
Comment (chMoussa) [reply docs/measurements/index.md]: Returning now a dictionary with two keys for this case

**Comment by chMoussa** [comment]
Thanks for the feedback @RolandMacDoland, made the requested few changes.

File: docs/measurements/index.md [hunk 12]
```diff
@@ -47,18 +47,28 @@ model = QuantumModel(
...
 tomo_options = {"n_shots": 100000}
-tomo_measurement = Measurements(protocol=Measurements.TOMOGRAPHY, options=tomo_options)
+tomo_measurement = Measurements(protocol=MeasurementProtocols.TOMOGRAPHY, options=tomo_options)
 
 # Get the exact expectation value.
 exact_values = model.expectation()
 
 # Run the tomography experiment.
-estimated_values_tomo = tomo_measurement(model)
+estimated_values_tomo = tomo_measurement(model, return_expectations=True)
```
Comment (RolandMacDoland) [inline docs/measurements/index.md]: Is there a need to pass on this argument ?

File: docs/measurements/index.md [hunk 13]
```diff
@@ -47,18 +47,28 @@ model = QuantumModel(
...
 ```
 
+## Getting measurements
+
+If we are interested in accessing the measurements for computing different quantities of interest other than the expectation values, we can access the measurement data via the `manager` attribute or simply pass `return_expectations=False` (a new round of measurements is performed and returned directly instead of computing expectation values) as follows:
+
+```python exec="on" source="material-block" session="measurements" result="json"
+
+measurements_tomo = tomo_measurement.manager.data
+print(measurements_tomo)
```
Comment (RolandMacDoland) [inline docs/measurements/index.md]: Maybe you can hide it ?

File: docs/measurements/index.md [hunk 14]
```diff
@@ -85,23 +95,59 @@ from qadence_protocols.measurements.utils_shadow import number_of_samples
...
 print(f"Estimated expectation value shadow = {estimated_values_shadow}") # markdown-exec: hide
 ```
 
+## Getting shadows
+
+If we are interested in accessing the measurement data from shadows, we can access the measurement data via the `manager` attribute or simply pass `return_expectations=False` (a new round of measurements is performed and returned directly instead of computing expectation values) as follows:
+
+```python exec="on" source="material-block" session="measurements" result="json"
+
+measurements_shadows = shadow_measurement.manager.data
```
Comment (RolandMacDoland) [inline docs/measurements/index.md]: Why not turning `MeasurementData` into a dataclass that holds `unitaries` and `measurements` as fields ? It'd be simpler than a key/value access imho.

File: docs/measurements/index.md (hunk 3 above)
Comment (RolandMacDoland) [reply docs/measurements/index.md]: See above.

File: docs/measurements/index.md (hunk 14 above)
Comment (RolandMacDoland) [reply docs/measurements/index.md]: Also: would it make sense to make the manager transparent ? Ie. make user exposed fields directly accessible in `Measurements` ?

File: qadence_protocols/measurements/abstract.py [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def validate_options(self) -> dict:
+        """Validate options passed to procotol and return a dict with default validated values."""
+        raise NotImplementedError
+
+
+class ShadowManagerAbstract(MeasurementManager, ABC):
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/abstract.py]:
```suggestion
class ShadowManager(MeasurementManager):
```

File: qadence_protocols/measurements/protocols.py [hunk 16]
```diff
@@ -1,115 +1,69 @@
...
 
     def __call__(
         self,
         model: QuantumModel,
         param_values: dict[str, Tensor] = dict(),
+        state: Tensor | None = None,
         return_expectations: bool = True,
     ) -> Tensor:
-        """Compute expectation values via measurements.
+        """Compute measurements or expectation values via measurements.
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/protocols.py]: Does it compute measurements ?

File: qadence_protocols/measurements/protocols.py [hunk 17]
```diff
@@ -1,115 +1,69 @@
...
-            module = importlib.import_module(PROTOCOL_TO_MODULE[self.protocol])
-        except (KeyError, ModuleNotFoundError, ImportError) as e:
-            raise type(e)(f"Failed to import Mitigations due to {e}.")
 
         conv_observables = model._observable
         observables = [obs.abstract for obs in conv_observables]
 
         # Partially pass the options and observable.
-        compute_fn = "compute_expectation" if return_expectations else "compute_measurements"
+        compute_fn = "expectation" if return_expectations else "measure"
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/protocols.py]: hm, it feels like there are two distinct return types here. For the sake of simplicity, I would choose one or the other but not let a method potentially return either as it makes its use quite confusing. That could be another PR.

File: qadence_protocols/measurements/shadow.py (hunk 9 above)
Comment (RolandMacDoland) [reply qadence_protocols/measurements/shadow.py:161]: What I meant is I didn't see where the `verify_options` is triggered before coming here. Maybe I just missed it.

File: qadence_protocols/measurements/tomography.py (hunk 10 above)
Comment (RolandMacDoland) [reply qadence_protocols/measurements/tomography.py]: Then you can remove them ?

File: qadence_protocols/measurements/protocols.py [hunk 18]
```diff
@@ -1,115 +1,69 @@
...
+    """
 
     def __init__(self, protocol: str, options: dict = dict()) -> None:
-        verified_options = MeasurementOptions(protocol, options).verify_options()
+        try:
+            module = importlib.import_module(PROTOCOL_TO_MODULE[protocol][0])
+            proto_class = getattr(module, PROTOCOL_TO_MODULE[protocol][1])
+        except (KeyError, ModuleNotFoundError, ImportError) as e:
+            raise type(e)(f"Failed to import Mitigations due to {e}.")
+        self.manager: MeasurementManager = proto_class(data=None, options=options)
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/protocols.py]: I am a bit confused: isn't the role of the `MeasurementManager` to _produce_ `MeasurementData` through for instance the `measure` method ? If so, why passing `data` as `None` to the initialiser ?

File: qadence_protocols/measurements/robust_shadow.py [hunk 19]
```diff
@@ -1,101 +1,159 @@
...
+    robust_local_shadow,
+    shadow_samples,
+)
+from qadence_protocols.types import MeasurementData
+
+
+class RobustShadowManager(ShadowManagerAbstract):
+    """The class for managing randomized robust shadow."""
+
+    def __init__(self, data: MeasurementData | None = None, options: dict = dict()):
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/robust_shadow.py]: Instead of data, would it make sense to pass `model`, `param_values` and `state` here ?

**Review by RolandMacDoland — COMMENTED** [review]
Thanks @chMoussa. I think we can go even further. The addition of a `MeasurementManager` is nice but I would somehow surface the functionalities directly to `Measurements` to make it transparent. Also, I am a bit dubious about passing `data` to it.

File: docs/measurements/index.md (hunk 12 above)
Comment (chMoussa) [reply docs/measurements/index.md]: No, just showing it for clarity

File: qadence_protocols/measurements/protocols.py (hunk 18 above)
Comment (chMoussa) [reply qadence_protocols/measurements/protocols.py]: The role of the manager is not only to produce `MeasurementData` but also provide extra functionalities to postprocess it. In principle, one could obtain data from somewhere else, and feed it as an input, but here None means no data was obtained yet.

Comment (RolandMacDoland) [reply qadence_protocols/measurements/protocols.py]: Indeed, but in practice, there is no place where you can externally pass `data` ? `Measurements` doesn't provide it and I am not sure `MeasurementManager` is exposed to the user ?

Comment (chMoussa) [reply qadence_protocols/measurements/protocols.py]: This is used for instance in testing faster that robust shadows are equal to the vanilla shadows when no calibration has been done.

Comment (RolandMacDoland) [reply qadence_protocols/measurements/protocols.py]: hm could you point me where ?

File: qadence_protocols/measurements/robust_shadow.py (hunk 19 above)
Comment (chMoussa) [reply qadence_protocols/measurements/robust_shadow.py]: Hm not really because I could be interested in testing other observables that are not defined within the `model`. One can think of reconstructing a state, and computing other quantities of interest out of it.

File: qadence_protocols/measurements/protocols.py (hunk 18 above)
Comment (chMoussa) [reply qadence_protocols/measurements/protocols.py]: No right, the current way was to do:
`robust_shadows.manager.data = shadow_measurements.manager.data`

I propably should set a property to access and a setter for data.

Comment (RolandMacDoland) [reply qadence_protocols/measurements/protocols.py]: 🤔

File: qadence_protocols/measurements/robust_shadow.py (hunk 19 above)
Comment (RolandMacDoland) [reply qadence_protocols/measurements/robust_shadow.py]: Alright, I think I see. What you mean is that you can post process the model by reconstructing the state using another observable.

File: qadence_protocols/measurements/shadow.py (hunk 8 above)
Comment (chMoussa) [reply qadence_protocols/measurements/shadow.py]: Removed `number_of_samples_and_medians`

File: qadence_protocols/types.py (hunk 11 above)
Comment (chMoussa) [reply qadence_protocols/types.py]: Eventually, `MeasurementData` becomes a dataclass

File: docs/measurements/index.md (hunk 12 above)
Comment (chMoussa) [reply docs/measurements/index.md]: Removed completely eventually

File: qadence_protocols/measurements/abstract.py (hunk 15 above)
Comment (chMoussa) [reply qadence_protocols/measurements/abstract.py]: Removed eventually.

File: qadence_protocols/measurements/protocols.py (hunk 17 above)
Comment (chMoussa) [reply qadence_protocols/measurements/protocols.py]: Removed

File: qadence_protocols/measurements/protocols.py (hunk 18 above)
Comment (chMoussa) [reply qadence_protocols/measurements/protocols.py]: Using a setter now with more transparency in Measurements

**Comment by chMoussa** [comment]
Hi @RolandMacDoland, eventually I change the `Measurements` so that they can be initialized and more bound to the input model, states, and parameter values as it makes more sense (note for tomography, the observables should be provided to be defined, but for a shadow not as the observable only plays a role in the expectation - yet it can be provided directly). We are also more transparent with the manager by allowing accessing several methods from `Measurements` directly. 

Each manager has now a `validate_data` method for checking that the data input is valid when passed by the user, given the definition of the protocol via its inputs.

Also we use now a dataclass `MeasurementData` with two fields.

File: qadence_protocols/measurements/protocols.py (hunk 16 above)
Comment (chMoussa) [reply qadence_protocols/measurements/protocols.py]: Fixed

File: docs/measurements/index.md [hunk 20]
```diff
@@ -47,18 +47,28 @@ model = QuantumModel(
 
 # Define a measurement protocol by passing the shot budget as an option.
 tomo_options = {"n_shots": 100000}
-tomo_measurement = Measurements(protocol=Measurements.TOMOGRAPHY, options=tomo_options)
+tomo_measurement = Measurements(protocol=MeasurementProtocols.TOMOGRAPHY, model=model, options=tomo_options)
```
Comment (RolandMacDoland) [inline docs/measurements/index.md]: I think we discussed to leave it as it was ?

File: docs/measurements/index.md [hunk 21]
```diff
@@ -47,18 +47,28 @@ model = QuantumModel(
...
-estimated_values_tomo = tomo_measurement(model)
+estimated_values_tomo = tomo_measurement()
 
 print(f"Exact expectation value = {exact_values}") # markdown-exec: hide
 print(f"Estimated expectation value tomo = {estimated_values_tomo}") # markdown-exec: hide
 ```
 
+## Getting measurements
+
+If we are interested in accessing the measurements for computing different quantities of interest other than the expectation values, we can access the measurement data via `data` as follows:
```
Comment (RolandMacDoland) [inline docs/measurements/index.md:64]: Nice :)

File: docs/measurements/index.md [hunk 22]
```diff
@@ -85,23 +95,55 @@ from qadence_protocols.measurements.utils_shadow import number_of_samples
...
 # Run the shadow experiment.
-estimated_values_shadow = shadow_measurement(model)
+estimated_values_shadow = shadow_measurement()
 
 print(f"Estimated expectation value shadow = {estimated_values_shadow}") # markdown-exec: hide
 ```
 
+## Getting shadows
+
+If we are interested in accessing the measurement data from shadows, we can access the measurement data via the `manager` attribute or simply pass `return_expectations=False` (a new round of measurements is performed and returned directly instead of computing expectation values) as follows:
```
Comment (RolandMacDoland) [inline docs/measurements/index.md]: It feels a bit contradictory with  the example below ?

File: docs/measurements/index.md [hunk 23]
```diff
@@ -110,29 +152,19 @@ model = QuantumModel(
...
 
 calibration = zero_state_calibration(N, n_qubits=2, n_shots=100, backend=model.backend, noise=noise)
 # This linear transformation should give us the probability error
-print(0.5 * (3.0 * calibration + 1))
+print(0.5 * (3.0 * calibration + 1)) # markdown-exec: hide
 
-Rshadow_options = {"shadow_size": N, "shadow_groups": K, "calibration": calibration}
-robust_shadow_measurement = Measurements(protocol=Measurements.ROBUST_SHADOW, options=Rshadow_options)
-estimated_values_robust_shadow = robust_shadow_measurement(model)
+Rshadow_options = {"shadow_size": N, "shadow_medians": K, "calibration": calibration}
```
Comment (RolandMacDoland) [inline docs/measurements/index.md]:
```suggestion
robust_shadow_options = {"shadow_size": N, "shadow_medians": K, "calibration": calibration}
```

File: pyproject.toml [hunk 24]
```diff
@@ -30,7 +30,8 @@ classifiers=[
 ]
 dependencies = [
     "torch",
-    "qadence[pulser, horqrux]==1.7.8",
+    "qadence[pulser, horqrux]",
```
Comment (RolandMacDoland) [inline pyproject.toml]: Why Horqrux btw ?

File: qadence_protocols/measurements/abstract.py [hunk 25]
```diff
@@ -0,0 +1,108 @@
...
+from abc import ABC, abstractmethod
+
+from qadence import QuantumModel
+from qadence.blocks.abstract import AbstractBlock
+from torch import Tensor
+
+from qadence_protocols.types import MeasurementData
+
+
+class MeasurementManager(ABC):
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/abstract.py:12]: Maybe turn it into a dataclass ? To avoid the initialisation boilerplate.

File: qadence_protocols/measurements/protocols.py [hunk 26]
```diff
@@ -1,115 +1,112 @@
...
-        return self.OPTIONS_TO_VERIFY[self.protocol]()
+    def __init__(
+        self,
+        protocol: str,
+        model: QuantumModel,
+        options: dict = dict(),
+        observables: list[AbstractBlock] = list(),
+        param_values: dict[str, Tensor] = dict(),
+        state: Tensor | None = None,
+        data: MeasurementData = MeasurementData(),
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/protocols.py]: I guess `model`, `observables`, `param_values` and `states` can be moved to `__call__`. `data` whenever it comes to experimental data can be passed in `options` ?

File: qadence_protocols/measurements/protocols.py [hunk 27]
```diff
@@ -1,115 +1,112 @@
...
-        if shadow_size is None and accuracy is None:
-            raise KeyError(
-                "Shadow protocol requires either an option"
-                " 'shadow_size' of type 'int' or 'accuracy' of type 'float'."
-            )
-        confidence = self.options.get("confidence", None)
-        if confidence is None:
-            raise KeyError("Shadow protocol requires an option 'confidence' of type 'float'.")
+        # Note that options and data are validated inside the manager
+        self._manager: MeasurementManager = proto_class(
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/protocols.py:45]: In `__call__` ?

File: qadence_protocols/measurements/protocols.py [hunk 28]
```diff
@@ -1,115 +1,112 @@
...
+            options, model, observables, param_values, state, data
+        )
 
-        return {"shadow_size": shadow_size, "accuracy": accuracy, "confidence": confidence}
+    @property
+    def data(self) -> MeasurementData:
+        return self._manager.data
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/protocols.py:51]: Nice.

File: qadence_protocols/measurements/protocols.py [hunk 29]
```diff
@@ -1,115 +1,112 @@
...
 
-    def verify_tomography(self) -> dict:
-        n_shots = self.options.get("n_shots")
-        if n_shots is None:
-            raise KeyError("Tomography protocol requires a 'n_shots' kwarg of type 'int').")
+    @data.setter
+    def data(self, newdata: MeasurementData) -> None:
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/protocols.py]:
```suggestion
    def data(self, new_data: MeasurementData) -> None:
```

File: qadence_protocols/measurements/protocols.py [hunk 30]
```diff
@@ -1,115 +1,112 @@
...
+        self,
+    ) -> Tensor:
+        """Reconstruct the state.
 
-        return {
-            "shadow_size": shadow_size,
-            "shadow_groups": shadow_groups,
-            "calibration": calibration,
-        }
+        Used only in shadow protocols.
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/protocols.py:107]: Maybe add a TODO about implementation in tomo ?

File: qadence_protocols/types.py [hunk 31]
```diff
@@ -14,6 +18,14 @@ def list(cls) -> list[str]:
         return list(map(lambda c: c.value, cls))  # type: ignore
 
 
+@dataclass
+class MeasurementData:
+    measurements: Tensor | list[Counter] | None = None
+    """Measurements data."""
+    unitaries: Tensor | None = None
```
Comment (RolandMacDoland) [inline qadence_protocols/types.py]: I'd default them to empty tensors or empty lists. it'll avoid the handling of the `None` case I think.

Comment (RolandMacDoland) [reply qadence_protocols/types.py]: Just wondering if the measurements field shouldn't be renamed `samples` and typed as a `list[Counter]` only.

File: tests/test_shadow.py [hunk 32]
```diff
@@ -223,22 +193,43 @@ def test_estimations_comparison_tomo_forward_pass(
...
+        model=model,
+        param_values=values,
+        options=robust_options,
+    )
 
-    robust_options = {"shadow_size": 54400, "shadow_groups": 6, "robust_correlations": None}
-    robust_shadows = Measurements(protocol=Measurements.ROBUST_SHADOW, options=robust_options)
-    robust_estimated_exp_shadow = robust_shadows(model, param_values=values)
+    # set measurement same as classical shadows
+    robust_shadows.data = shadow_measurements.data
```
Comment (RolandMacDoland) [inline tests/test_shadow.py:211]: Oh wow so nice :)

File: qadence_protocols/mitigations/analog_zne.py [hunk 33]
```diff
@@ -159,14 +160,14 @@ def analog_zne(
...
 ) -> Tensor:
-    if noise is None or noise.protocol not in supported_noise_models:
-        if model._noise is None or model._noise.protocol not in supported_noise_models:
+    if noise is None or noise.protocol[-1] not in supported_noise_models:
+        if model._noise is None or model._noise.protocol[-1] not in supported_noise_models:
             raise ValueError(
-                "A Noise.DEPOLARIZING or Noise.DEPHASING model must be provided"
-                " either to .mitigate() or through the <class QuantumModel>."
+                "A NoiseProtocol.ANALOG noise model must be provided"
+                + " either to .mitigate() or through the <class QuantumModel>."
```
Comment (RolandMacDoland) [inline qadence_protocols/mitigations/analog_zne.py]:
```suggestion
                " either to .mitigate() or through the <class QuantumModel>."
```

**Review by RolandMacDoland — COMMENTED** [review]
Thanks a lot @chMoussa. Getting there really.

File: qadence_protocols/mitigations/readout.py [hunk 34]
```diff
@@ -262,13 +273,15 @@ def mitigation_minimization(
 def mitigate(
     model: QuantumModel,
     options: dict,
-    noise: Noise | None = None,
+    noise: NoiseHandler | None = None,
     param_values: dict[str, Tensor] = dict(),
 ) -> list[Counter]:
-    if noise is None or noise.protocol != Noise.READOUT:
-        if model._noise is None or model._noise.protocol != Noise.READOUT:
+    if noise is None or (not isinstance(noise.protocol[-1], NoiseProtocol.READOUT)):
```
Comment (RolandMacDoland) [inline qadence_protocols/mitigations/readout.py]: Do you have to use `isinstance` here ? if `noise.protocol` is expected to be a list of `NoiseProtocol` it shouldn't be a problem to compare them ?

File: docs/measurements/index.md (hunk 20 above)
Comment (chMoussa) [reply docs/measurements/index.md]: Yeah but eventually, the protocols depend on these inputs, so better changing it.

File: pyproject.toml (hunk 24 above)
Comment (chMoussa) [reply pyproject.toml]: Hm probably an old test that is not here anymore was removed. Will change this soon

File: qadence_protocols/mitigations/readout.py (hunk 34 above)
Comment (chMoussa) [reply qadence_protocols/mitigations/readout.py]: Using filter now

File: docs/measurements/index.md (hunk 22 above)
Comment (chMoussa) [reply docs/measurements/index.md]: Ah my bad, fixed

File: qadence_protocols/types.py (hunk 31 above)
Comment (chMoussa) [reply qadence_protocols/types.py]: Hm I can do but tensors are better suited for the shadows IMO.

Comment (chMoussa) [reply qadence_protocols/types.py]: Removed None

File: qadence_protocols/measurements/protocols.py (hunk 27 above)
Comment (RolandMacDoland) [reply qadence_protocols/measurements/protocols.py:45]: I think this makes for a more meaningful UX, ie:

```python
measurement  = Measurement(protocol, options) <-- setting the measurement protocol. experiemental data can be passed in options. 
measurement(model, ...) <-- applying to a model with necessary options. It also makes the measurement reusable to various models.
```

**Review by RolandMacDoland — COMMENTED** [review]
Hey @chMoussa, last one from me i promise. Otherwise happy to approve.

**Comment by chMoussa** [comment]
> Hey @chMoussa, last one from me i promise. Otherwise happy to approve.

@RolandMacDoland I now have model set in __call__ and other functions besides `reconstruct_state` where I force to having set a model previously.

File: docs/measurements/index.md [hunk 35]
```diff
@@ -47,18 +47,28 @@ model = QuantumModel(
 
 # Define a measurement protocol by passing the shot budget as an option.
 tomo_options = {"n_shots": 100000}
-tomo_measurement = Measurements(protocol=Measurements.TOMOGRAPHY, options=tomo_options)
+tomo_measurement = Measurements(protocol=MeasurementProtocols.TOMOGRAPHY, options=tomo_options)
```
Comment (RolandMacDoland) [inline docs/measurements/index.md]:
```suggestion
tomo_measurement = Measurements(protocol=MeasurementProtocol.TOMOGRAPHY, options=tomo_options)
```

I would singularise.

File: qadence_protocols/measurements/abstract.py (hunk 25 above)
Comment (RolandMacDoland) [reply qadence_protocols/measurements/abstract.py:12]: Any opinion on this ?

**Review by RolandMacDoland — APPROVED** [review]
Thanks a million @chMoussa that's really nice work. Minor quibbles if you wish to address them. Otherwise, LGTM.
